include/ada/url_pattern.h

```diff
@@ -0,0 +1,323 @@
+
+  std::string_view get_delimiter() const ada_warn_unused;
+  std::string_view get_prefix() const ada_warn_unused;
+
+  // @see https://urlpattern.spec.whatwg.org/#options-ignore-case
+  bool ignore_case = false;
+
+  static url_pattern_compile_component_options DEFAULT;
+  static url_pattern_compile_component_options HOSTNAME;
+  static url_pattern_compile_component_options PATHNAME;
+
+ private:
+  // @see https://urlpattern.spec.whatwg.org/#options-delimiter-code-point
+  std::optional<char> delimiter{};
+  // @see https://urlpattern.spec.whatwg.org/#options-prefix-code-point
+  std::optional<char> prefix{};
+};
+
+// A struct providing the URLPattern matching results for a single
+// URL component. The URLPatternComponentResult is only ever used
+// as a member attribute of a URLPatternResult struct. The
+// URLPatternComponentResult API is defined as part of the URLPattern
+// specification.
+struct url_pattern_component_result {
+  std::string input;
+  std::unordered_map<std::string, std::string> groups;
+};
+
+class url_pattern_component {
+ public:
+  url_pattern_component() = default;
+
+  // This function explicitly takes a std::string because it is moved.
+  // To avoid unnecessary copy, move each value while calling the constructor.
+  url_pattern_component(std::string&& new_pattern, std::regex&& new_regexp,
+                        std::regex_constants::syntax_option_type&& new_flags,
+                        std::vector<std::string>&& new_group_name_list,
+                        bool new_has_regexp_groups)
+      : pattern(std::move(new_pattern)),
+        flags(new_flags),
+        regexp(std::move(new_regexp)),
+        group_name_list(std::move(new_group_name_list)),
+        has_regexp_groups_(new_has_regexp_groups) {}
+
+  // @see https://urlpattern.spec.whatwg.org/#compile-a-component
+  template <url_pattern_encoding_callback F>
+  static tl::expected<url_pattern_component, url_pattern_errors> compile(
+      std::string_view input, F encoding_callback,
+      url_pattern_compile_component_options& options);
+
+  // @see https://urlpattern.spec.whatwg.org/#create-a-component-match-result
+  url_pattern_component_result create_component_match_result(
+      std::string_view input, const std::smatch& exec_result);
+
+  std::string_view get_pattern() const noexcept ada_lifetime_bound
+      ada_warn_unused;
+  const std::regex& get_regexp() const noexcept ada_lifetime_bound
+      ada_warn_unused;
+  std::regex_constants::syntax_option_type get_regexp_flags() const noexcept
+      ada_lifetime_bound ada_warn_unused;
+  const std::vector<std::string>& get_group_name_list() const noexcept
+      ada_lifetime_bound ada_warn_unused;
+  bool has_regexp_groups() const noexcept ada_lifetime_bound ada_warn_unused;
+
+  std::string to_string() const;
+
+ private:
+  std::string pattern{};
+  std::regex_constants::syntax_option_type flags = std::regex::ECMAScript;
+  std::regex regexp{"*"};
```

Comment:
```suggestion
  std::regex regexp{".*"};
```

here I saw that wpt_urlpattern_tests.has_regexp_groups test passes when we use .* for ecmascript compatibility

Reply:
yes, nice catch.